app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RetrofitProvider.kt

```diff
@@ -0,0 +1,21 @@
+package org.codeforvictory.android.recipes.data.api.recipe
+
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+
+object RetrofitProvider {
+
+    private var singleton: Retrofit? = null
+    val value: Retrofit
```

Comment:
should we do full double check locking?

Reply:
It's been a really long time since I heard about that technique. I could make changes but I guess there a few things to have in consideration:

1. Since we are always creating the dependency when initializing objects (at construction time), multithreading shouldn't be a big issue, should it?

2. I've read a few things about the pattern and it doesn't seem to be recommended as a general practice (in favor of other lazy init approaches). See for reference:

- https://www.baeldung.com/java-singleton-double-checked-locking
- https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Could we get more context on why this would be needed?

Reply:
There are alternatives, but double check locking is pretty good for a few situations.

I guess that what @Mr-Gibbs means is that we are trying to do it thread-safe by using `synchronized`, but actually it is not.

Reply:
If we wanna use lazy, we can use lazy {} ?

Reply:
+1 for `lazy` or even just relying on the lazy initialization of `object`s and directly go:

```
val value: Retrofit = Retrofit.Builder()
  .baseUrl("https://www.themealdb.com/api/json/v1/1")
  .addConverterFactory(MoshiConverterFactory.create())
  .build()
```

Reply:
> If we wanna use lazy, we can use lazy {} ?

Both look good to me! I had actually thought about using `lazy` in the first place. I tried to create our own delegate but it was too verbose for what we needed. Thanks for bringing this up guys!

Actually, I looked at the implementation of `lazy` again (it's been a while), and I saw they actually used the full double check locking pattern you were talking about, right?

```
...
if (_v1 !== UNINITIALIZED_VALUE) {
      @Suppress("UNCHECKED_CAST")
      return _v1 as T
  }

  return synchronized(lock) {
      val _v2 = _value
      if (_v2 !== UNINITIALIZED_VALUE) {
          @Suppress("UNCHECKED_CAST") (_v2 as T)
          ...
  ...
```

Reply:
Addressed: [`4606ac9` (#13)](https://github.com/CodeForVictory/recipes-mvi-playground/pull/13/commits/4606ac9ce1e6d7155cba6ed2f1f3145ed6ee3929)